airrflow bytesize transcript (#1414)
nf-core/nf-co.re: markdown/events/2022/bytesize_nf-core-airrflow.md

Typically for sequencing, Miseq sequencers are used, because they allow for longer read lengths, that covers the complete BDJ and beginning of the C region.

[5:39](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=339)
How is the bioinformatic analysis done for these kind of sequences? This typically does not happen like a traditional RNA-Seq analysis and that's because mapping to a reference genome is challenging in this case, due to the high diversity of these BCR and TCR sequences. Also, it's a highly repetitive genome region with all of these BD and J gene segments there. For this kind of analysis, specific tools are used and the reads are aligned to specific reference data, also for these BCR and TCR receptors.

[6:15](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=375)

> I always say luckily for us, when wanting to write a pipeline to do this kind of analysis, there's already plenty of tools out there that can analyze this data. One of the better well-known ones is the immcantation framework that is developed by the Kleinstein lab in Yale. It's an open source toolset to analyze AIRRseq data from beginning to end. There's a whole community of users already using this framework and here you can also get the details, in case you want to have more information.

Reviewer: ```suggestion
I always say luckily for us, when wanting to write a pipeline to do this kind of analysis, there's already plenty of tools out there that can analyze this data. One of the better well-known ones is the Immcantation framework that is developed by the Kleinstein lab in Yale. It's an open-source toolset to analyze AIRRseq data from beginning to end. There's a whole community of users already using this framework and here you can also get the details, in case you want to have more information.
```

> Thanks to developing this pipeline as part of the nf-core community, we gained visibility and we quickly found quite some collaborators to develop the pipeline. I want to mention that this is really a community-based development effort! Susanna Marquez from the immcantation lab joined early in the beginning and also David Ladd from Monash University helped by adding some features. Whereas initially, Alex Peltzer,

Reviewer: ```suggestion
Thanks to developing this pipeline as part of the nf-core community, we gained visibility and we quickly found quite some collaborators to develop the pipeline. I want to mention that this is really a community-based development effort! Susanna Marquez from the Immcantation lab joined early in the beginning and also David Ladd from Monash University helped by adding some features. Whereas initially, Alex Peltzer,
```

Afterwards, there's a step for clonal analysis, which identifies which of the sequences of the BCR belong to which B-cell clone. It assigns the individual BCR sequences and TCR sequences to their specific clone and in the case of the B-cells it can also perform lineage reconstruction of the whole B-cell clone. Finally there's a step for reporting, doing repertoire analysis and reporting, including QC reports by MultiQC. That's the general steps of the pipeline, but now if we look a bit more detailed there is a ton more processes - individual processes that are part of the pipeline. I'm going to explain them a bit in more detail now.

[9:01](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=541)

> Starting for the QCN sequence assembly, the pipeline supports different sequencing protocols, including multiplex PCR, in which the users have to provide the B and C-primer sequences that were used for the amplification, or in the case of 5' RACE, providing the C-primer and the linker sequences for amplification. Both protocols are supported with and without UMI barcodes. The barcodes can also be provided in different configurations. Starting from the raw sequencing data, a sample sheet needs to be provided that contains sample information and the individual FASTQ files for all samples. Depending on if the sequencing protocol includes UMI barcodes or not, there will be some processes but they all start with quality control of the reads with FASTQ, filtering the sequences by quality threshold, masking the primer sequences and if a UMI-based protocol is used, then a consensus is built from all the sequences that have the same UMI barcode. This way, it also allows to correct for the errors as I mentioned before.

Reviewer: ```suggestion
Starting for the QC and sequence assembly, the pipeline supports different sequencing protocols, including multiplex PCR, in which the users have to provide the V and C-primer sequences that were used for the amplification, or in the case of 5' RACE, providing the C-primer and the linker sequences for amplification. Both protocols are supported with and without UMI barcodes. The barcodes can also be provided in different configurations. Starting from the raw sequencing data, a sample sheet needs to be provided that contains sample information and the individual FASTQ files for all samples. Depending on if the sequencing protocol includes UMI barcodes or not, there will be some processes but they all start with quality control of the reads with FastQC, filtering the sequences by quality threshold, masking the primer sequences and if a UMI-based protocol is used, then a consensus is built from all the sequences that have the same UMI barcode. This way, it also allows to correct for the errors as I mentioned before.
```

There is an extra procedure, that is employed whenever it's estimated that the length of the UMI barcodes will not be sufficient to cover all of the diversity of the sample and that is bypassed by first clustering all the sequences by similarity and annotating the cluster ID and then two different sequences with the same UMI barcode can also be distinguished this way.

[10:51](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=651)

> After building the consensus, all the sequences that contain the same UMI barcode are collapsed and the count, the number of sequences with the same UMI barcode, is annotated. Other metadata is also annotated as well as the count of duplicate sequences with different UMI barcodes that were collapsed. This can be useful for filtering. After this step there comes the VBJ assignment and filtering step and here it's also possible

Reviewer: ```suggestion
After building the consensus, all the sequences that contain the same UMI barcode are collapsed and the count, the number of sequences with the same UMI barcode, is annotated. Other metadata is also annotated as well as the count of duplicate sequences with different UMI barcodes that were collapsed. This can be useful for filtering. After this step there comes the VDJ assignment and filtering step and here it's also possible
```

> to start with already assembled sequences, that can be provided with a sample sheet and FASTA files. Typically single cell sequencing data processing starts at this step, because the pipeline supports directly the output from the tool 10x Genomics CellRanger multi, which provides, while incorporating also TCR and VCR sequencing in the 10x Genomics sequencing procedure, the output of that tool, which is the AIRR rearrangement

Reviewer: ```suggestion
to start with already assembled sequences, that can be provided with a sample sheet and FASTA files. Typically single cell sequencing data processing starts at this step, because the pipeline supports directly the output from the tool 10x Genomics CellRanger multi, which provides, while incorporating also TCR and BCR sequencing in the 10x Genomics sequencing procedure, the output of that tool, which is the AIRR rearrangement
```

> table, thatt contains all of the sequences there. This can also be directly provided to the pipeline at that step. What that step does, it aligns the sequences to the IMGT reference and then assigns what exact B, D and J segments were used there. In the case of the single cell data there is an option. This gene reassignment step is optional.

Reviewer: ```suggestion
table, that contains all of the sequences there. This can also be directly provided to the pipeline at that step. What that step does, it aligns the sequences to the IMGT reference and then assigns what exact V, D and J segments were used there. In the case of the single cell data there is an option. This gene reassignment step is optional.
```

[12:29](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=749)
After alignment to the IMGT reference there is a number of quality filtering steps that are performed. First it's checked that the locus matches exactly the V_calls. The V(D)J segment assignment, that there are a minimum of 200 informative positions, maximum 10% and nucleotides, that the sequences that are determined are productive, that the junction region is a multiple of three amino acids. There's also a possibility of removing chimeric reads, detecting contamination across samples and finally collapsing duplicate sequences if there are any. Plenty of quality filters as part of the pipeline.

[13:18](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=978)

> The next step is clonal analysis. In that case hierarchical clustering is used based on the hamming distances between sequences. The pipeline is also able to auto-detect a hamming distance threshold that can be used to determine, which sequences are part of the same clone, or are part of different clones. There is also a step for lineage reconstruction of the clonal lineage trees. Recently the pipeline offers to use the Enchanter tool, which is developed by Susanna Marcus in immcantation. That tool provides calls to other immcantation tools and also nice reports for each of these steps. I invite you to check them out here.

Reviewer: ```suggestion
The next step is clonal analysis. In that case hierarchical clustering is used based on the hamming distances between sequences. The pipeline is also able to auto-detect a hamming distance threshold that can be used to determine, which sequences are part of the same clone, or are part of different clones. There is also a step for lineage reconstruction of the clonal lineage trees. Recently the pipeline offers to use the EnchantR tool, which is developed by Susanna Marquez in Immcantation. That tool provides calls to other Immcantation tools and also nice reports for each of these steps. I invite you to check them out here.
```

[14:43](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=883)
Here we will see an example of this repertoire analysis report. First there is a summary of all the samples used for the analysis. Clonal abundance and clonal diversity are reported, together with vision usage. Finally all of the tools that are used as part of the pipeline and their citations are noted here, to make it really easy for users of the pipeline to also cite the original tools, that are being used.

[15:12](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=912)
As you know, all documentation for nf-core/airrflow can be found on the nf-core website, so check it out if you want to use it. There are also some example results of the pipeline when the full tests are run on AWS.

[15:30](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=930)

> What's next for this pipeline? Stay tuned for a new release that comes really soon - we hope this week or the next. It will includes more quality control and reporting as part of the Enchanter tool, as I have mentioned, by Susanna Marquez, as well as and code refactoring using subworkflows.

Reviewer: ```suggestion
What's next for this pipeline? Stay tuned for a new release that comes really soon - we hope this week or the next. It will includes more quality control and reporting as part of the EnchantR tool, as I have mentioned, by Susanna Marquez, as well as and code refactoring using subworkflows.
```

[15:50](https://www.youtube.com/watch?v=CrJgxVRVlqY&t=950)

> At this point I would like to thank all the contributors to the pipeline: Simon Heumos and myself at QBiC, Alex Peltzer who was initially at QBiC but now is at Böhringer Ingelheim, Susanna Marquez at the Kleinstein Lab in Yale, David Ladd at Monash University and some collaborations also at the University of Tübingen, Christophe Rochel and Marcus Kovaric.

Reviewer: ```suggestion
At this point I would like to thank all the contributors to the pipeline: Simon Heumos and myself at QBiC, Alex Peltzer who was initially at QBiC but now is at Böhringer Ingelheim, Susanna Marquez at the Kleinstein Lab in Yale, David Ladd at Monash University and some collaborations also at the University of Tübingen, Christoph Ruschil and Markus Kowaric.
```

> If you have any questions don't hesitate to join us, join the airrflow channel on nf-core Slack and if you have any questions related to the immcantation tools, you also have the contact emails to contact them directly.

Reviewer: ```suggestion
If you have any questions don't hesitate to join us, join the airrflow channel on nf-core Slack and if you have any questions related to the Immcantation tools, you also have the contact emails to contact them directly.
```
